Refactorings + generic return types + proper Jackson configuration + more

martinzima opened the pull request:
Initially, I just wanted to fix the deserialization to support methods with generic return types, but then I stumbled across more issues and ended up with doing some refactorings to the core and with a few other improvements/bugfixes, so I'll be happy to get it included in the main branch. :-)

Please review the last bugfix I mentioned in the commit description, i.e. the Params.asFormEncodedRequestBody method. For me, it seems almost like if the request body was returned not-URL-encoded on purpose until now, but AFAIK, that is not correct and should be the same as with the query string encoding. But maybe I'm just overlooking something and missing the point of this method?

martinzima committed:
Merge pull request #1 from mmazi/develop

Merge from base

martinzima committed:
-Refactored lot of internal classes and tests (e.g. better split JSON de/serialization from network code)
-Added support for methods returning generic types
-Jackson is now configured from one place (effective for both params and return values)
-Fixed bug of POST form-urlencoded data not encoded + other minor improvements

mmazi commented:
Wow, that's one large commit, thanks a lot! I'd like to review it before accepting it though. Unfortunately, I won't have time to do it in the next two weeks. Are you in a hurry with this getting merged?

Also, if this solves several issues, I'd really prefer if each issue was in its own commit if this is possible. Thanks!

martinzima commented:
No need to hurry, I have no problem with living on my own fork until then, take your time for a proper review. :-)

I tried to shortly describe the addressed issues in the commit description, but since I ended up with refactoring multiple things eventually, I'm afraid I won't be able to divide it into separate commits now.

mmazi commented:
@reddragcz , I've reviewed your pull request and merged it. This is a wonderful contribution, thanks a lot!

In case you want to contribute any further changes in the future, I'd be glad to give you commit rights to the project.
